packages/framework/src/Model/Product/Search/FilterQuery.php

```diff
@@ -237,6 +237,63 @@ public function getMaxCurrentSellingPriceRuntimeField(int $pricingGroupId): arra
         ];
     }
 
+    /**
+     * @param int $pricingGroupId
+     * @return array[]
+     */
+    protected function getHasActiveSpecialPriceRuntimeField(int $pricingGroupId): array
+    {
+        $script = "
+        boolean hasActiveSpecialPrice = false;
+        double basicPrice = 0;
+        DateTimeFormatter formatter = java.time.format.DateTimeFormatter.ofPattern('yyyy-MM-dd HH:mm:ss').withZone(java.time.ZoneOffset.UTC);
+        
+        if (!params['_source']['prices'].isEmpty()) {
+            for (def price : params['_source']['prices']) {
+                if (price['pricing_group_id'] === params['pricing_group_id']) {
+                    basicPrice = price['price_with_vat'];
+                    break;
+                }
+            }
+        }
+
+        if (!params['_source']['special_prices'].isEmpty()) {
+            def currentDate = java.time.ZonedDateTime.parse(params['current_date'], formatter).toInstant();
+
+            for (def specialPrice : params['_source']['special_prices']) {
+                def validFrom = java.time.ZonedDateTime.parse(specialPrice['valid_from'], formatter).toInstant();
+                def validTo = java.time.ZonedDateTime.parse(specialPrice['valid_to'], formatter).toInstant();
+
+                if ((validFrom.isBefore(currentDate) || validFrom.equals(currentDate)) &&
+                    (validTo.isAfter(currentDate) || validTo.equals(currentDate))) {
+
+                    for (def specialPriceData : specialPrice['prices']) {
+                        if (basicPrice > specialPriceData['price_with_vat']) {
+                            hasActiveSpecialPrice = true;
+                            break;
```

Comment:
I had trouble with variants. 

can you please double-check this scenario?

- set category "Food" as auto-filtered by special price only
- assign "Television Philips [M]" (id: 69) to the "Food" category
- set special price in price list "Items on sale" (id: 3) for variant 54" Philips CRT [V] (catnum: "9176588")
- see the main variant is not displayed in the food category on storefront

Reply:
nice catch, fixed in 5a9b41a7eb